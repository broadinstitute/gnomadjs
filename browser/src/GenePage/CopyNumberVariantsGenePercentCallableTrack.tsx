import React from 'react'

import { referenceGenome } from '@gnomad/dataset-metadata/metadata'
// import CoverageTrack from '../CoverageTrack'
import Query from '../Query'
import DiscreteBarPlot from '../DiscreteBarPlot'

type OwnProps = {
  datasetId: string
  chrom: string
  start: number
  stop: number
}

// @ts-expect-error TS(2456) FIXME: Type alias 'Props' circularly references itself.
type Props = OwnProps & typeof CopyNumberVariantsGenePercentCallableTrack.defaultProps

// @ts-expect-error TS(7022) FIXME: 'CopyNumberVariantsGenePercentCallableTrack' implicitly has type '... Remove this comment to see the full error message
const CopyNumberVariantsGenePercentCallableTrack = ({ datasetId, geneId }: Props) => {
  const operationName = 'CopyNumberVariantsGenePercentCallableTrack'
  const query = `
  query ${operationName}($geneId: String!, $datasetId: CopyNumberVariantDatasetId!, $referenceGenome: ReferenceGenomeId!) {
    gene(gene_id: $geneId, reference_genome: $referenceGenome) {
            cnv_track_callable_coverage(dataset: $datasetId) {
                xpos
                percent_callable
                position
                contig
            }
            start
            stop
            chrom
        }
    }
`
  return (
    <Query
      operationName={operationName}
      query={query}
      variables={{ geneId, datasetId, referenceGenome: referenceGenome(datasetId) }}
      loadingMessage="Loading coverage"
      loadingPlaceholderHeight={220}
      errorMessage="Unable to load coverage"
      success={(data: any) => {
        if (!data.gene || !data.gene.cnv_track_callable_coverage) {
          return false
        }
        return data.gene && data.gene.cnv_track_callable_coverage
      }}
    >
      {({ data }: any) => {
        const transformedArray = data.gene.cnv_track_callable_coverage.map((item: any) => ({
          pos: item.position,
          percent_callable: item.percent_callable,
          xpos: item.xpos,
        }))
        transformedArray.sort((a: any, b: any) => a.xpos - b.xpos)

        const coverage = [
          {
            color: 'rgb(70, 130, 180)',
            buckets: transformedArray,
            name: 'percent callable',
            opacity: 0.7,
          },
        ]
        const geneStart = data.gene.start
        const geneStop = data.gene.stop
        const geneChrom = Number(data.gene.chrom)

        return (
          <DiscreteBarPlot
            datasets={coverage}
            height={200}
            regionStart={geneStart}
            regionStop={geneStop}
            chrom={geneChrom}
          />
        )
      }}
    </Query>
  )
}

export default CopyNumberVariantsGenePercentCallableTrack
