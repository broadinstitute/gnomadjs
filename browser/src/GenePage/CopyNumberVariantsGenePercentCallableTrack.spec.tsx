import React from 'react'
import { createRenderer } from 'react-test-renderer/shallow'

import { jest, describe, expect, test } from '@jest/globals'
import { mockQueries } from '../../../tests/__helpers__/queries'
import Query, { BaseQuery } from '../Query'

import CopyNumberVariantsGenePercentCallableTrack from './CopyNumberVariantsGenePercentCallableTrack'

import { allDatasetIds, hasCopyNumberVariantCoverage } from '@gnomad/dataset-metadata/metadata'
import geneFactory from '../__factories__/Gene'

jest.mock('../Query', () => {
  const originalModule = jest.requireActual('../Query')

  return {
    __esModule: true,
    ...(originalModule as object),
    default: jest.fn(),
    BaseQuery: jest.fn(),
  }
})

const { resetMockApiCalls, resetMockApiResponses, simulateApiResponse, setMockApiResponses } =
  mockQueries()

beforeEach(() => {
  Query.mockImplementation(
    jest.fn(({ query, children, operationName, variables }) =>
      simulateApiResponse('Query', query, children, operationName, variables)
    )
  )
  ;(BaseQuery as any).mockImplementation(
    jest.fn(({ query, children, operationName, variables }) =>
      simulateApiResponse('BaseQuery', query, children, operationName, variables)
    )
  )
})

afterEach(() => {
  resetMockApiCalls()
  resetMockApiResponses()
})

const datasetsWithCoverage = allDatasetIds.filter((datasetId) =>
  hasCopyNumberVariantCoverage(datasetId)
)

describe.each(datasetsWithCoverage)(
  'CopyNumberVariantsGenePercentCallableTrack with dataset %s',
  (datasetId) => {
    test('queries with appropriate params', () => {
        const gene = geneFactory.build()
        setMockApiResponses({
        CopyNumberVariantsGenePercentCallableTrack: () => ({
          gene: {
            cnv_track_callable_coverage: [],
          },
        }),
      })

      const shallowRenderer = createRenderer()
      shallowRenderer.render(
        <CopyNumberVariantsGenePercentCallableTrack
          datasetId={datasetId}
          geneId={gene.gene_id}
        />
      )

      expect(shallowRenderer.getRenderOutput()).toMatchSnapshot()
    })
  }
)
