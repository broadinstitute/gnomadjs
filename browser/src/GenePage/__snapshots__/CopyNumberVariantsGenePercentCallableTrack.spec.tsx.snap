// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyNumberVariantsGenePercentCallableTrack with dataset gnomad_cnv_r4 queries with appropriate params 1`] = `
<mockConstructor
  errorMessage="Unable to load coverage"
  loadingMessage="Loading coverage"
  loadingPlaceholderHeight={220}
  operationName="CopyNumberVariantsGenePercentCallableTrack"
  query="
  query CopyNumberVariantsGenePercentCallableTrack($geneId: String!, $datasetId: CopyNumberVariantDatasetId!, $referenceGenome: ReferenceGenomeId!) {
    gene(gene_id: $geneId, reference_genome: $referenceGenome) {
            cnv_track_callable_coverage(dataset: $datasetId) {
                xpos
                percent_callable
                position
                contig
            }
            start
            stop
            chrom
        }
    }
"
  success={[Function]}
  variables={
    {
      "datasetId": "gnomad_cnv_r4",
      "geneId": "dummy_gene-1",
      "referenceGenome": "GRCh38",
    }
  }
>
  [Function]
</mockConstructor>
`;
